### ISSUE TITLE
Page reload in Maui Hybrid Blazor app causes unhandled exception

### ISSUE THREAD

**Issue by crui3er** [dotnet/aspnetcore#52119]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

BlazorWebView can reload page by itself. E.g. it might happen when app is taken foreground on Android after a while.
When page is reloading, WebViewManager will try to dispose previois PageContext.
https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebView/WebView/src/WebViewManager.cs#L221
PageContext disposing will causes disposing on components of the app.
Components disposing can produce calls to JS. E.g. to dispose related js components.
But Page has already a new state, old Page state is lost.
E.g. if we invoke js on IJSObjectReference, it will fail with JSException with message like `'blazor.webview.js:1  JS object instance with ID 1 does not exist (has it been disposed?).'`. 
Normally app can't handle such exception and Page (already reloaded) will notified about unhandled exception and will stop working.
![image](https://github.com/dotnet/aspnetcore/assets/20703715/b83677ae-cdef-4d2c-aa73-5fc162162b49)

This issue happens only when there is resolved scoped service that implements `IAsyncDisposable` which returns not completed task.
In this situation JS invoke from component dispose method could be dispatched before new page context is created.

This issue is similar to https://github.com/dotnet/aspnetcore/pull/32901

Proposed solution.
1) Mark PageContext as Disposed on calling PageContext .DisposeAsync
https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebView/WebView/src/PageContext.cs#L57
2) Mark linked WebViewJSRuntime as Disconnected. Disconnected WebViewJSRuntime should throw JSDisconnectedException on attemt to invoke JS like RemoteJSRuntime does it.
3) IpcSender. We should have special instance of IpcSender which can be linked with PageContext and do nothing if linked PageContext is disposed. This special instance on IpcSender should be passed to services in PageContext constructor (WebViewNavigationManager, WebViewJSRuntime, WebViewRenderer). This way we will protect new state of the page in webview from corruption.
4) IpcReceiver. It should also ignore received message if related PageContext is disposed. Such messages processing starts at https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebView/WebView/src/IpcReceiver.cs#L49

I faced this issue in MAUI HybridBlazor project but since all related code is located in this repo, I reported the issue here.

### Steps To Reproduce

https://github.com/crui3er/MauiHybridBlazorReloadIssue
Run app on Windows or Android (I did not test on other platfoms).
Click 'Simulate page reload' button.
Unhandled exception happened.

### .NET Version

dotnet 7

**Comment by mkArtakMSFT** [comment]
@crui3er did you try to run this project with .NET 8 and see if you run into the same issue?

**Comment by crui3er** [comment]
@mkArtakMSFT yes, it's reproducible on .net 8. I'just tested it on Windows app.
Also I updated repo with sample project.

**Comment by pigwing** [comment]
I have the same problem.

**Comment by kimdiego2098** [comment]
+1

**Comment by eltrixo** [comment]
Same problem here, only started after .NET 8 upgrade and it happens on all our pages that have QuickGrids.

I have reproduced in a vanilla .NET MAUI Blazor Hybrid app with QuickGrid and the simplest "Getting started" sample code.

**Comment by kubo990** [comment]
+1

**Comment by ghood97** [comment]
+1

**Comment by efonsecab** [comment]
I'm having a similar issue when trying to force reload after authentication in a MAUI Hybrid app.
![image](https://github.com/dotnet/aspnetcore/assets/3481899/5b3d020f-8799-4730-aa9a-e9c1f0d18c82)

This is what the DevTools are showing
![image](https://github.com/dotnet/aspnetcore/assets/3481899/6bbcce2a-7d54-4798-828c-aec6b1cb68f4)

**Comment by LukaszJuroszek** [comment]
.net 9.0.0-preview.7.24405 - the same issue

**Comment by dhruvindudhat** [comment]
Is there any workaround for this issue?

**Comment by Lindsay-Mathieson** [comment]
> Is there any workaround for this issue?

Disable Page reload by key

```c#
blazorWebView.BlazorWebViewInitialized += Bwv_Initialized;

        private async void Bwv_Initialized(object? sender, BlazorWebViewInitializedEventArgs e)
        {
            await blazorWebView.WebView.EnsureCoreWebView2Async();
            blazorWebView.WebView.CoreWebView2.Settings.AreBrowserAcceleratorKeysEnabled = false;
        }

```

**Comment by dhruvindudhat** [comment]
@Lindsay-Mathieson I already disabled refresh using F5 in release mode using this code but when application under long process and we kept machine in ideal state for some time then this error occurs at bottom in yellow background panel with refresh option and application lost it's state.

**Comment by martinbu** [comment]
Hiding all elements before a reload occures fixes the issue for me. See here:
https://github.com/microsoft/fluentui-blazor/issues/1036#issuecomment-2309420135

**Comment by dhruvindudhat** [comment]
@martinbu  In my case it's auto reloading and page shown exception panel at bottom in application. so how hiding the element works when page auto reload occurs?

**Comment by dhruvindudhat** [comment]
@mkArtakMSFT @javiercn  Can you suggest any workaround for this issue? I can not wait to fix this issue in upcoming releases. I have upcoming release and this issue is frequently happening in blazor maui hybrid app[for mac catalyst]. If you have any idea please suggest.here is detail explanation.

Exception : 
**Exceptions were encountered while disposing components. (JS object instance with ID 6 does not exist (has it been disposed?).
S@app://0.0.0.0/_framework/blazor.webview.js:1:2468
@app://0.0.0.0/_framework/blazor.webview.js:1:3326
Promise@[native code]
beginInvokeJSFromDotNet@app://0.0.0.0/_framework/blazor.webview.js:1:3317
@app://0.0.0.0/_framework/blazor.webview.js:1:40716
@user-script:1:4:75**

Do you know why this exception occurs in blazor hybrid app? In my case this is happens only for mac catalyst and windows looks perfect!

**Comment by Robula** [comment]
Same issue for us using `Microsoft.AspNetCore.Components.WebView.WindowsForms.BlazorWebView` in WinForms on SDK 8.0.400.

**Comment by dhruvindudhat** [comment]
Is there any way to just prevent the auto reload/refresh in maui blazor app for mac catalyst?

**Comment by Dunge** [comment]
> > Is there any workaround for this issue?
> 
> Disable Page reload by key
> 
> blazorWebView.BlazorWebViewInitialized += Bwv_Initialized;
> 
>         private async void Bwv_Initialized(object? sender, BlazorWebViewInitializedEventArgs e)
>         {
>             await blazorWebView.WebView.EnsureCoreWebView2Async();
>             blazorWebView.WebView.CoreWebView2.Settings.AreBrowserAcceleratorKeysEnabled = false;
>         }

Using the latest .NET9 and MAUI versions I do not seem to have a "WebView" property under BlazorWebView.

I'm using the BlazorFluentUI library which seems to be related. I'm getting this error randomly when starting my app, not really forcing a refresh or disposing specific components. Anyone have a better catch-all workaround solution?

**Comment by true-perfect-code** [comment]
In a MAUI Blazor App NET 9 I use FluentUI Blazor and had the problem constantly when logging out. I have integrated logout via MainLayout-Page and there I have several Fluent components which seem to not release JS resources (that's why we get the problem).
My solution was pretty simple, I created a bool variable in the codebehind
`OpenMainLayout=true) `
and generated the Razor Page depending on the variable 
`@if (OpenMainLayout))`
When the user logs out, I set the 
`OpenMainLayout=false, `
execute 
```
StateHasChanged(); 
Task.Delay(300); 
```
and after that I make a redirect via 
`NavigationManager!.NavigateTo(“/”, true);`

This seems to work well, I no longer get a JS error message in the console and the app is redirected to the root page (which was not the case before).
